app/controllers/admin/pages_controller.rb

```diff
@@ -96,5 +93,31 @@ def prepopulate_new_form(slug)
                 Page.new
               end
     end
+
+    def update_and_overwrite_landing_page
+      if page_params["overwrite"] == "true"
+        Page.transaction do
+          current_landing_page = Page.find_by(landing_page: true)
+          current_landing_page&.update(landing_page: false)
+
+          @page.update(page_params)
+        end
```

Comment:
**blocking**: I guess the purpose of having this transaction is to either save both or neither BUT:

1. the controller action already runs in a transaction
2. the first update is going to trigger https://github.com/forem/forem/blob/8d632d14478a3ca14fb09d4b8fd6f567e03945cf/app/models/page.rb#L13 which runs asynchronously, so there's no "rolling back" that. 

I have a suggestion which gets rid of the `overwrite` param and moves the data integrtiy control back to the model object instead of having it here in the controller: an `after_commit` callback.

The page is updated with the classic `.update(page_params)`, the callback checks if `landing_page` was turned from false to true by checking [previous_changes](https://api.rubyonrails.org/classes/ActiveModel/Dirty.html#method-i-previous_changes). If that's changed, you do something like: `Page.where.not(id: id).update_all(landing_page: false)`. [`update_all()`](https://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-update_all) bypasses all callbacks as it's a straight SQL call to the DB

This way, wherever that page is updated from, you don't run the risk of having multiple columns with `landing_page` true.

That said, you would lose the validation BUT I also see you're not using it for anything but showing the modal, so we might just simply show the modal if a landing page exists...

Let me know if any of this isn't clear, I'm happy to help 🙌

Reply:
I very much support this idea! We could stil combine it with the service object I suggested further up or do on its own, I'm fine with either.

Reply:
Hey @citizen428, I changed my mind a bit. I agree in principle that all that I have suggested is correct but I paired with Julianna on this PR, both for the client side and the server side ends of this functionality and I think we should hold off on refactoring it because the feature itself works and this idea of mine is going to add quite a bit to her workload and it's not strictly necessary as we reached a functioning state for the UX in our paired session.

This is what I'm suggesting:

- @juliannatetreault is going to finalize the PR plumbing in the missing functionality on the client side to get everything working
- we categorize both the service object idea and the restructure I proposed as a refactoring ticket/issue (basically copy pasting all of these suggestions in the description there)
- we review and merge this keeping that in mind
- I'll pair again with @juliannatetreault (or whomever wants to work on the refactoring issue) on a separate PR, with a fresh mind and the knowledge that the feature is fullfilling its purpose

There's not going to be a long while between this PR and the refactoring PR so we don't run the risk of forgetting how this feature works.

For this reason, I'm going to remove the `blocking` comment from my suggestion. Let me know if you're okay with this approach

Reply:
Yes, 100% okay with this approach, my comments weren't necessarily meant to be addressed immediately.